### PR TITLE
[RNMobile] Adjust vertical margins in InnerBlock

### PR DESCRIPTION
**Pull request by jbinda**
## Description
<!-- Please describe what you have changed or added -->
PR provides additional login to adjust vertical margins when navigate through InnerBlock hierarchy.

@iamthomasbishop,
With this changes navigate inside InnerBlock becomes smoother. The spaces between block is kept with the same value. Please taka a look on presented GIFs

Please also refer to:
[Issue](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1789)
[Related gutenberg-mobile PR](https://github.com/wordpress-mobile/gutenberg-mobile/pull/1838)

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

1. Add below code in `initial-html`. It provides some nested InnerBlock structure

<details>

```
/**
 * @format
 * @flow
 */

export default `
<!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:heading -->
<h2>Test Heading</h2>
<!-- /wp:heading --></div></div>
<!-- /wp:group -->

<!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"></div></div>
<!-- /wp:group -->

<!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group -->

<!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:image -->
<figure class="wp-block-image"><img alt=""/></figure>
<!-- /wp:image --></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group -->

<!-- wp:image -->
<figure class="wp-block-image"><img alt=""/></figure>
<!-- /wp:image --></div></div>
<!-- /wp:group -->

<!-- wp:image -->
<figure class="wp-block-image"><img alt=""/></figure>
<!-- /wp:image --></div></div>
<!-- /wp:group -->

<!-- wp:image -->
<figure class="wp-block-image"><img alt=""/></figure>
<!-- /wp:image --></div></div>
<!-- /wp:group -->

<!-- wp:image -->
<figure class="wp-block-image"><img alt=""/></figure>
<!-- /wp:image -->

<!-- wp:image -->
<figure class="wp-block-image"><img alt=""/></figure>
<!-- /wp:image --></div></div>
<!-- /wp:group -->


`;
```

</details>

2. Try to play with nested Group block and see that vertical margins between block should be kept with same value
3. Notice that there is no margins multiplication anymore when nest multiple empty Group Blocks
4. Feel free to complicate this structor further more and observe the vertical margin behaviour

## Screenshots <!-- if applicable -->

<img src="https://user-images.githubusercontent.com/21242757/73441404-3d6b2500-4353-11ea-8709-de75bb6a4be9.gif" width="300" />     <img src="https://user-images.githubusercontent.com/21242757/73441405-3d6b2500-4353-11ea-9d54-b2573aff5385.gif" width="300" />

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

Bug fix / refactor

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->.

### CODE REVIEW

**Comment by iamthomasbishop** [comment]
This looks great, well done!

I have a separate question: I noticed in some cases (particularly noted on the iOS screen cast above), multiple levels of children (dashed outlines) are shown. Example:

Screen Shot 2020-01-30 at 8 34 51 AM

I believe only the first child level should be visibly outlined.

**Comment by jbinda** [comment]
@iamthomasbishop 
The example you gave it's slight different than you described with below words:
>multiple levels of children (dashed outlines) are shown

It is default `AppenderButton` behaviour which instead of (+) Icon with solid border renders dashed borders as placeholder. 

In e.g. `Image` Block placeholder makes nice look with dashed border applied and light grey background. Comparing group placeholder to that you can be confused a little bit. By on the other hand it make sense and it's consistent with behaviour in the root list of blocks. Please see below GIF. There are two Groups in the root list. The first one have Heading and empty Group block inside. Second one is empty and do not have any nested blocks. As I mentioned above the double dashed border is because with InnerBlock styling logic we apply dashed border to all children of selected Group. We don't do the same at the RootList level.




>I believe only the first child level should be visibly outlined.

That's the way how it works but the effect you noticed is caused by reason I explained above.

I hope all above make sense. 

The possible solution that I have in my mind are:
1. Conditionally turn off applying the dashed border for child group that is empty (then the dashed border will comes from the placeholder)
2. Conditionally turn off dashed border **in group placeholder** when the group is inside InnerBlock
3. Instead of render placeholder render the `AppenderButton`  with some dimming of different style that point it's state to inactive
4. Add placeholder that is more similar to other blocks placeholder

Keep in mind the first two solution brings more complexity to bordering style in the navigation-dow logic because depending of placement in hierarchy we should modify the margins as well to align the content

Please let me know what do you think and how we should approach this according to open Group block to production.

**Comment by jbinda** [comment]
I'm not sure it's a blocker here.

However I also think that it might be worth to add some animation in `Breadcrums` on `FloatingToolbar` when selection inside InnerBlock changes. I thought initially about little dim out/ dim in the name/icon of blocks (or similar animation that we have after click "Go Up" button).

Why ?

E.g. because when you nest multiple empty Group block you can loose orientation after press (the UI doesn't change) but in fact the selection does and you go deeper and deeper after each press. In other cases it will bring some nice-looking effect. Please take a look on below GIF:

The structure of Blocks is presented here (nested 8 empty Group blocks):

<details>

```
<!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"><!-- wp:group -->
<div class="wp-block-group"><div class="wp-block-group__inner-container"></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group --></div></div>
<!-- /wp:group -->
```

</details>



Please observe that you do not see any effect when you select 3rd to 7th (yeah I'm pressing the block repeatedly). The UI changes only when you select the last one (because the placeholder is replaced with AppenderButton) and when you select first two (apply bordering style, breadcrumbs show the current hierarchy, navigation arrows is disabled). 

WDYT ?

File: packages/base-styles/_variables.scss [hunk 1]
```diff
@@ -76,6 +76,8 @@ $block-selected-to-content: $block-edge-to-content - $block-selected-margin - $b
 $block-selected-child-to-content: $block-selected-to-content - $block-selected-child-margin - $block-selected-child-border-width;
 $block-custom-appender-to-content: $block-selected-margin - $block-selected-border-width;
 $block-media-container-to-content: $block-selected-child-margin + $block-selected-border-width;
+$block-selected-vertical-margin-descendant: 3 * $block-selected-child-to-content;
+$block-selected-vertical-margin-child: 3 * $block-selected-child-to-content - 6 * $block-selected-child-border-width;
```
Comment (koke) [inline packages/base-styles/_variables.scss]: I don't understand these calculations, can you explain what's going on here?

File: packages/block-editor/src/components/block-list/block.native.js [hunk 2]
```diff
@@ -290,6 +296,8 @@ export default compose( [
 		const isInnerBlockHolder = name === getGroupingBlockName();
 		const isRootListInnerBlockHolder = ! isSelectedBlockNested && isInnerBlockHolder;
 
+		const shouldApplyVerticalMarginStyle = ! isLastBlock && ( ( isFirstBlock && parentCount === 2 ) || parentCount > 2 );
```
Comment (koke) [inline packages/block-editor/src/components/block-list/block.native.js]: If I understand this correctly, this is trying to add a margin to every sibling except the last one, right?

This expression seems more complicated than it needs, if I understand each variable correctly. I see three possible cases regarding the number of children in the parent: 

- There's one block (`parentCount < 2`). Then this must be the last (and first) block, so `! isLastBlock === false` and the margin shouldn't apply.
- There are two blocks (`parentCount == 2`). We have already set a condition to not add a margin on the last block, so the `isFirstBlock` is only relevant for the other case, which will always be true: if I'm not the last of 2, I must be the first.
- There are more than two blocks (`parentCount > 2`). Similarly, we've ruled out the case where we are the last block, and we want to add the margin for every other block.

So, if I'm not missing anything, the same logic should work with this simpler condition:

```suggestion
		const shouldApplyVerticalMarginStyle = ! isLastBlock;
```

File: packages/block-library/src/index.native.js [hunk 3]
```diff
@@ -148,7 +148,7 @@ export const registerCoreBlocks = () => {
 		mediaText,
 		preformatted,
 		gallery,
-		devOnly( group ),
+		group,
```
Comment (koke) [inline packages/block-library/src/index.native.js]: I know that we want to enable this on production, but this should go in its own PR

**Comment by koke** [comment]
> Conditionally turn off dashed border in group placeholder when the group is inside InnerBlock

I think that one makes the most sense. The placeholder border "conflicts" with the nested blocks child border, and should only be used in root-level groups.

File: packages/block-library/src/index.native.js (hunk 3 above)
Comment (jbinda) [reply packages/block-library/src/index.native.js]: it was added to this PR because of discussion on Slack but its not a problem to do this separately

File: packages/block-editor/src/components/block-list/block.native.js (hunk 2 above)
Comment (jbinda) [reply packages/block-editor/src/components/block-list/block.native.js]: I have check it and it's fine but after we modify `isLastBlock` as below:

<image src="https://user-images.githubusercontent.com/21242757/73541574-1ab03d80-4433-11ea-8d86-0e1e52a097c7.png" width="500" />

In previous implementation it tries to check the order base of current `clientId` InnerBlocks count. Which isn't proper because we need to check the order in parent InnerBlock list.

I have checked and it seems that this flag isn't use anywhere so probably we can also prevent to pass it to component props

**Comment by jbinda** [comment]
> I think that one makes the most sense. The placeholder border "conflicts" with the nested blocks child border, and should only be used in root-level groups.

Logically it works fine because if the group is empty it has a dashed placeholder. Considering UX  I think it also looks fine. 

However I do not have strong opinion on that and we can try implement above solution.

File: packages/block-editor/src/components/block-list/block.native.js (hunk 2 above)
Comment (jbinda) [reply packages/block-editor/src/components/block-list/block.native.js]: I have also simplify passing the prop in latest commit

File: packages/base-styles/_variables.scss [hunk 4]
```diff
@@ -76,6 +76,8 @@ $block-selected-to-content: $block-edge-to-content - $block-selected-margin - $b
 $block-selected-child-to-content: $block-selected-to-content - $block-selected-child-margin - $block-selected-child-border-width;
 $block-custom-appender-to-content: $block-selected-margin - $block-selected-border-width;
 $block-media-container-to-content: $block-selected-child-margin + $block-selected-border-width;
+$block-selected-vertical-margin-descendant: 2 * $block-selected-to-content;
+$block-selected-vertical-margin-child: $block-edge-to-content;
```
Comment (jbinda) [inline packages/base-styles/_variables.scss]: When block is a descendant each block gets bottom margin of double of `$block-selected-to-content`. Then it will equals the original space between blocks (it also includes padding applied by our logic when parent is selected).

Space between child is set to `$block-edge-to-content` because our bordering rule apply also some margin to each block (with `$block-selected-margin` and `$block-selected-border-width`)

So in each case space between block content should be equal.

File: packages/base-styles/_variables.scss (hunk 1 above)
Comment (jbinda) [reply packages/base-styles/_variables.scss]: I have answer to that after pushing changes because this one is out of date. I did a mistake implementing the logic here.

File: packages/base-styles/_variables.scss (hunk 4 above)
Comment (koke) [reply packages/base-styles/_variables.scss]: Ah, after testing this for a bit I think I understand why this works that way. The double space is so that the children stay in place after being selected:

Untitled 2

**Review by koke — APPROVED** [review]
It's a bit confusing navigating these many nested groups, but I think this is behaving correctly 😅

File: packages/base-styles/_variables.scss (hunk 4 above)
Comment (jbinda) [reply packages/base-styles/_variables.scss]: Yeah, with this there is less "jumpy-effect" when selection change.

I think after we implement ideal version of FloatingToolbar we can also reduce it even more.

You think above implementation make sense ?

**Comment by jbinda** [comment]
> It's a bit confusing navigating these many nested groups, but I think this is behaving correctly 😅

I think we also bring some refactor to improve UX according to [this comment](https://github.com/WordPress/gutenberg/pull/19960#issuecomment-580621561) according to [this convo](https://github.com/wordpress-mobile/gutenberg-mobile/issues/1802)


One more question. Shall we remove the `devOnly` flag with this PR or create a new one ?

File: packages/base-styles/_variables.scss (hunk 4 above)
Comment (koke) [reply packages/base-styles/_variables.scss]: Yeah, once I figured out what it was doing it made sense. It wasn't super clear just by looking at the variable names, but I don't have a better suggestion

**Comment by koke** [comment]
> One more question. Shall we remove the devOnly flag with this PR or create a new one ?

Ah, yes please, let's move that change to a different PR

File: packages/block-library/src/index.native.js (hunk 3 above)
Comment (jbinda) [reply packages/block-library/src/index.native.js]: done

File: packages/base-styles/_variables.scss (hunk 4 above)
Comment (jbinda) [reply packages/base-styles/_variables.scss]: ok, thanks !
